File: Problem2_ConstructBT.java
/************ Time Complexity: O(n) ************/
/************ Space Complexity: O(n) ************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        //base case
        if(preorder.length==0)
            return null;
        
        // considering 1st element of preorder to be root
        TreeNode root = new TreeNode(preorder[0]);
        //initialization
        int index = -1;
        for(int i=0; i<inorder.length; i++){
            if(inorder[i]==root.val){
                //root = i index
                index = i;
            }
        }
        
        // inorder_left = 0 -> index
        int[] inorder_left = Arrays.copyOfRange(inorder, 0, index);
        // inorder_right = index+1 -> inorder.length
        int[] inorder_right= Arrays.copyOfRange(inorder, index+1, inorder.length);
        // preorder_left = 1 -> index+1
        int[] preorder_left= Arrays.copyOfRange(preorder, 1, index+1);
        // preorder_right = index+1 -> preorder.length
        int[] preorder_right= Arrays.copyOfRange(preorder, index+1, preorder.length);
        
        root.left = buildTree(preorder_left, inorder_left);
        root.right = buildTree(preorder_right, inorder_right);
    
        return root;
    }
}

File: Problem2_ConstructBT.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    Dictionary<int, int> inorderMap = new Dictionary<int, int>();
    int idx = 0;
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        for(int i = 0; i < inorder.Length; i++) {
            this.inorderMap.Add(inorder[i],i);
        }
       
        return BuildNode(preorder, 0, inorder.Length - 1);


    }

    private TreeNode BuildNode(int[] preorder, int start, int end) {
        //base
        if(start > end) return null;

        int rootVal = preorder[idx];
        idx++;
        int rootIndex = inorderMap[rootVal];
        TreeNode root = new TreeNode(rootVal);
        root.left = BuildNode(preorder, start, rootIndex - 1);

        root.right = BuildNode(preorder, rootIndex + 1, end);
        return root;
    }
}